add handling canceled subscriptions

satococoa opened the pull request:
## Summary
Ensure proper handling of canceled subscriptions by checking subscription status.

## Changes

### Code
- Modified to check subscriptions only with `active` status.
- Refactored the route06 domain check for easier debugging (ideally, this should be factored out into an Environment value 🤔 ).

## TODO before merging this PR
On stripe's dashboard

- [x] `dev` on Sandbox environment: Enable receiving the `customer.subscription.updated` and `customer.subscription.deleted` event.
- [x]  Live environment: Enable receiving the `customer.subscription.updated` and `customer.subscription.deleted` event.

## Testing

### How to test in your local environment

#### Preparation
1. Add `NEXT_PUBLIC_SITE_URL="http://localhost:3000"` to .env.local (the port may vary on your environment)
2. If you use an email with the `route06.co.jp` domain, disable the skipping of the subscription check by editing the following code.
  - Using an alias is recommended: e.g., `satoshi.ebisawa+debug20241021-1@route06.co.jp`

```ts
diff --git a/lib/utils.ts b/lib/utils.ts
index 89e08f5..1c792a2 100644
--- a/lib/utils.ts
+++ b/lib/utils.ts
@@ -6,6 +6,7 @@ export function cn(...inputs: ClassValue[]) {
 }
 
 export const isEmailFromRoute06 = (email: string): boolean => {
+	return false;
 	const domain = email.split("@")[1];
 	return domain ? domain.endsWith("route06.co.jp") : false;
 };
```

4. Install the Stripe CLI and log in. Follow this instruction: https://docs.stripe.com/stripe-cli
5. Activate the local webhook listener: `stripe listen --skip-verify --forward-to localhost:3000/webhooks/stripe` (the port may vary)
  - The webhook signing secret will be printed out.
6. Start server: STRIPE_WEBHOOK_SECRET=[webhook signing secret] bun dev

#### Testing steps
1. Sign in
2. Purchase a subscription.
3. Go Payment Portal through `/settings/billing`
4. Cancel the subscription.
5. You'll receive `customer.subscription.update`: You **can** use Giselle at this point because the subscription is still `active`.
6. On stripe dashboard, use [test clocks](https://docs.stripe.com/billing/testing/test-clocks) to simulate the subscription end.
  - Instructions here: https://docs.stripe.com/billing/testing/test-clocks/simulate-subscriptions
8. You'll receive a `customer.subscription.deleted` event.
9. Now, you will be required to purchase a new subscription if you access giselle.
10. Finish test clocks' simulation.
  - ⚠️ Finishing a simulation deletes all objects associated with it, including the customer and subscription, from the sandbox or test mode. It doesn’t affect live mode. https://docs.stripe.com/billing/testing/test-clocks/simulate-subscriptions

----

### How to test in preview environment

#### Preparation
- ⚠️ You need to sign up with an email that is **not** from the `route06.co.jp` domain.
- Purchase a subscription.

#### A. Easy way: Edit the Database Directly

Locate your subscription in the `subscriptions` table and Change the status to `canceled`.

#### B. Complicated way: Use stripe's [test clocks](https://docs.stripe.com/billing/testing/test-clocks) feature

Register webhook on the preview URL to receive `customer.subscription.updated` and `customer.subscription.deleted`.
(I have not done this.)

1. Go Payment Portal through `/settings/billing`
2. Cancel the subscription. (You'll receive `customer.subscription.update` event on preview environment.)
3. On stripe dashboard, use [test clocks](https://docs.stripe.com/billing/testing/test-clocks) to simulate the subscription end.
4. You'll receive a `customer.subscription.deleted` event in preview environment.
5. Now, you will be required to purchase a new subscription if you access giselle.
6. Finish test clocks' simulation.
  - ⚠️ Finishing a simulation deletes all objects associated with it, including the customer and subscription, from the sandbox or test mode. It doesn’t affect live mode. https://docs.stripe.com/billing/testing/test-clocks/simulate-subscriptions

----

## Other Information

- [The Subscription Object](https://docs.stripe.com/api/subscriptions/object)
- [Subscription webhook events](https://docs.stripe.com/billing/subscriptions/webhooks#events)
- https://docs.stripe.com/webhooks

shige reviewed (approved):
LGTM 👍 

I confirmed that it works! ✅ 
Test clocks are very useful. 🕐 

---
```sh
STRIPE_WEBHOOK_SECRET=xxx bun dev

🔔  Webhook received: customer.updated
 POST /webhooks/stripe 400 in 255ms
🔔  Webhook received: test_helpers.test_clock.created
 POST /webhooks/stripe 400 in 237ms
🔔  Webhook received: customer.subscription.deleted
🔔  Webhook received: test_helpers.test_clock.ready
 POST /webhooks/stripe 400 in 228ms
 POST /webhooks/stripe 200 in 1747ms
```

satococoa commented on services/accounts/actions.ts:
```diff
@@ -1 +1 @@
-export { retrieveStripeSubscriptionBySupabaseUserId } from "./actions/retrieve-stripe-subscription";
+export { retrieveActiveStripeSubscriptionBySupabaseUserId as retrieveStripeSubscriptionBySupabaseUserId } from "./actions/retrieve-stripe-subscription";
```
Sorry, this was a mistake. I used the editor's rename symbol function, but it was not intended.

satococoa replied:
fixed in [e0d8679](https://github.com/route06inc/giselle/pull/34/commits/e0d86795ce55d6b17527603cdcd8d854acead970)

satococoa commented:
@toyamarinyon 

> Is it correct that this is already implemented in app/webhooks/stripe/route.ts?"

Yes. handling webhook is already implemented.
However, the updated Stripe subscription statuses were not used.

satococoa commented:
Thank you for the review! 
I'll complete the 'TODO before merging this PR' tasks and merge it tomorrow. 🚀

satococoa commented:
> dev on Sandbox environment: Enable receiving the customer.subscription.updated and customer.subscription.deleted event.

<img width="293" alt="スクリーンショット 2024-10-23 8 46 18" src="https://github.com/user-attachments/assets/70e7c040-c899-42cd-8879-1d5d543319e5">

> Live environment: Enable receiving the customer.subscription.updated and customer.subscription.deleted event.

<img width="585" alt="スクリーンショット 2024-10-23 8 48 36" src="https://github.com/user-attachments/assets/4aa1887d-4e32-425e-a7ab-da7958c26acb">

satococoa commented:
✅ received webhooks
<img width="1213" alt="スクリーンショット 2024-10-23 9 33 41" src="https://github.com/user-attachments/assets/42e122bb-fd0a-4965-be67-644056d903dc">

I confirmed that a user who has canceled (deleted) their subscription will be redirected to the /subscriptions/checkout page.
(However, the attempt failed due to the absence of  `NEXT_PUBLIC_SITE_URL` environmental variables. )

Since it is working correctly, the operation check is complete 🚀
